src/databricks/labs/lsql/dashboards.py

```diff
@@ -37,111 +51,612 @@
+            "--filter",
+            type=str,
+            action="extend",
+            dest="filters",
+            nargs="*",
+        )
+        parser.add_argument(
+            "--overrides",
+            type=json.loads,
+            help="Override the low-level Lakeview entities with a json payload.",
+        )
+        return parser
 
-    def as_dict(self) -> dict[str, str]:
-        return dataclasses.asdict(self)
+    def _parse_header(self, header: str) -> dict:
+        """Header is an argparse string."""
+        header_split = shlex.split(header)
+        parser = self._get_arguments_parser()
+        try:
+            return vars(parser.parse_args(header_split))
+        except (argparse.ArgumentError, SystemExit) as e:
+            logger.warning(f"Parsing {self._path}: {e}")
+            return vars(parser.parse_known_args(header_split)[0])
 
+    def split(self) -> tuple[str, str]:
+        """Split the query file header from the contents.
 
-class WidgetMetadata:
-    def __init__(
+        The optional header is the first comment at the top of the file.
+        """
+        try:
+            parsed_query = sqlglot.parse_one(self._content, dialect=sqlglot.dialects.Databricks)
+        except sqlglot.ParseError as e:
+            logger.warning(f"Parsing {self._path}: {e}")
+            return "", self._content
+
+        if parsed_query.comments is None or len(parsed_query.comments) == 0:
+            return "", self._content
+
+        first_comment = parsed_query.comments[0]
+        return first_comment.strip(), self._content
+
+
+class MarkdownHandler(BaseHandler):
+    """Handle Markdown files."""
+
+    _FRONT_MATTER_BOUNDARY = re.compile(r"^-{3,}\s*$", re.MULTILINE)
+
+    def _parse_header(self, header: str) -> dict:
+        """Markdown configuration header is a YAML."""
+        _ = self
+        return yaml.safe_load(header) or {}
+
+    def split(self) -> tuple[str, str]:
+        """Split the markdown file header from the contents.
+
+        The header is enclosed by a horizontal line marked with three dashes '---'.
+        """
+        splits = self._FRONT_MATTER_BOUNDARY.split(self._content, 2)
+        if len(splits) == 3:
+            _, header, content = splits
+            return header.strip(), content.lstrip()
+        if len(splits) == 2:
+            logger.warning(f"Parsing {self._path}: Missing closing header boundary.")
+        return "", self._content
+
+
+@unique
+class WidgetType(str, Enum):
+    """The query widget type"""
+
+    AUTO = "AUTO"
+    TABLE = "TABLE"
+    COUNTER = "COUNTER"
+
+    def as_widget_spec(self) -> type[WidgetSpec]:
+        widget_spec_mapping: dict[str, type[WidgetSpec]] = {
+            "TABLE": TableV1Spec,
+            "COUNTER": CounterSpec,
+        }
+        if self.name not in widget_spec_mapping:
+            raise ValueError(f"Can not convert to widget spec: {self}")
+        return widget_spec_mapping[self.name]
+
+
+class TileMetadata:
```

Comment:
let's make it a dataclass, so that we avoid `copy.deepcopy` and the whole class of bugs associated with that.